### PR TITLE
[Merged by Bors] - refactor(CategoryTheory): the structure Functor.FullyFaithful

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces a structure containing the data involved in a fully faithful functor `F`, i.e. inverse maps to `F.map`. (This was suggested by Andrew Yang.)

This should also restore certain definitional properties which may have been lost in #12449 and #12462.

Better names have been given to lemmas `IsIso.of_iso` and `IsIso.of_iso_inv` which are now `Iso.isIso_hom` and `Iso.isIso_inv`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by YaelDillies** [comment]
I moved
> This was suggested by @erdOne.

to under `---` to avoid Andrew getting pinged on Github whenever someone merges the future commit on master into their branch.

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean [hunk 1]
```diff
@@ -127,113 +129,108 @@ theorem preimageIso_mapIso (f : X ≅ Y) : F.preimageIso (F.mapIso f) = f := by
...
 @[simps]
-noncomputable def natTransOfCompFullyFaithful (α : F ⋙ H ⟶ G ⋙ H) :
-    F ⟶ G where
-  app X := (equivOfFullyFaithful H).symm (α.app X)
-  naturality X Y f := by
-    dsimp
-    apply H.map_injective
-    simpa using α.naturality f
-#align category_theory.nat_trans_of_comp_fully_faithful CategoryTheory.natTransOfCompFullyFaithful
-#align category_theory.nat_trans_of_comp_fully_faithful_app CategoryTheory.natTransOfCompFullyFaithful_app
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/Functor/FullyFaithful.lean:186]: It follows from `Functor.FullyFaithful.whiskeringRight` in `CategoryTheory.Whiskering`. It is used at https://github.com/leanprover-community/mathlib4/pull/12929/files#diff-8957fa6b8f07ce56865d16fe111fc970905093796767be98db633fe1a3cde4bbR139-R140

[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/CategoryTheory/Functor/FullyFaithful.lean:186]: Ah I didn't see that while skimming through the diff.

**Comment by erdOne** [comment]
Oops, forgot about this.
Thanks a lot!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**